<?xml version="1.0" encoding="utf-8"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="binocular_laser_profile_scanner">
  
  <link name="world"/>

  <joint name="world_joint" type="fixed">
      <parent link="world"/>
      <child link="base_link"/>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
  </joint>

  <link
    name="base_link">
    <inertial>
      <origin
        xyz="0.12801 9.7736E-11 0.01878"
        rpy="0 0 0" />
      <mass
        value="4.0349" />
      <inertia
        ixx="0.032847"
        ixy="-3.6828E-13"
        ixz="-0.00198"
        iyy="0.058337"
        iyz="2.1579E-12"
        izz="0.090143" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/base_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/base_link.STL" />
      </geometry>
    </collision>
  </link>
  <link
    name="rot_table_link">
    <inertial>
      <origin
        xyz="1.1102E-16 -2.2551E-17 -0.0055825"
        rpy="0 0 0" />
      <mass
        value="0.72806" />
      <inertia
        ixx="0.003993"
        ixy="-9.6296E-35"
        ixz="5.7977E-20"
        iyy="0.003993"
        iyz="6.229E-36"
        izz="0.0079545" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/rot_table_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/rot_table_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="rot_table_joint"
    type="revolute">
    <origin
      xyz="0 0 0.04"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="rot_table_link" />
    <axis
      xyz="0 0 -1" />
    <limit
      lower="0"
      upper="6.284"
      effort="1000"
      velocity="1" />
  </joint>
  <link
    name="motor_link">
    <inertial>
      <origin
        xyz="-1.7241E-05 0.094802 -0.070622"
        rpy="0 0 0" />
      <mass
        value="0.032668" />
      <inertia
        ixx="0.00019299"
        ixy="3.5712E-09"
        ixz="-1.2204E-10"
        iyy="9.3754E-05"
        iyz="8.587E-05"
        izz="9.9894E-05" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/motor_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.25098 0.25098 0.25098 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/motor_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="motor_joint"
    type="revolute">
    <origin
      xyz="0.18 0 0.0682"
      rpy="0 1.5708 0" />
    <parent
      link="base_link" />
    <child
      link="motor_link" />
    <axis
      xyz="0 0 1" />
    <limit
      lower="0"
      upper="3.142"
      effort="1000"
      velocity="1" />
  </joint>
  <link
    name="scanner_body_link">
    <inertial>
      <origin
        xyz="5.5511E-17 0.018028 0.0081343"
        rpy="0 0 0" />
      <mass
        value="0.075035" />
      <inertia
        ixx="3.2489E-05"
        ixy="8.8797E-20"
        ixz="-3.3005E-19"
        iyy="0.00027821"
        iyz="-6.8601E-09"
        izz="0.00027056" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/scanner_body_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/scanner_body_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="scanner_body_joint"
    type="fixed">
    <origin
      xyz="-0.004 0.18003 -0.18005"
      rpy="1.5708 -1.5708 0" />
    <parent
      link="motor_link" />
    <child
      link="scanner_body_link" />
    <axis
      xyz="0 0 0" />
  </joint>
  <link
    name="line_laser_link">
    <inertial>
      <origin
        xyz="-5.5511E-17 0 -0.01825"
        rpy="0 0 0" />
      <mass
        value="0.0040188" />
      <inertia
        ixx="4.8224E-07"
        ixy="-2.6518E-23"
        ixz="1.079E-22"
        iyy="4.8224E-07"
        iyz="6.4388E-24"
        izz="7.1033E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/line_laser_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/line_laser_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="line_laser_joint"
    type="fixed">
    <origin
      xyz="0 0.0175 0.026246"
      rpy="0 0 0" />
    <parent
      link="scanner_body_link" />
    <child
      link="line_laser_link" />
    <axis
      xyz="0 0 0" />
  </joint>
  <link
    name="left_camera_link">
    <inertial>
      <origin
        xyz="-2.7718E-06 -8.7642E-07 0.0071874"
        rpy="0 0 0" />
      <mass
        value="0.0016829" />
      <inertia
        ixx="5.6994E-08"
        ixy="-6.0636E-12"
        ixz="-5.3509E-12"
        iyy="5.6998E-08"
        iyz="2.5605E-12"
        izz="2.9435E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/left_camera_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/left_camera_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="left_camera_joint"
    type="fixed">
    <origin
      xyz="0.073979 0.0175 0.031514"
      rpy="0 -0.85197 0" />
    <parent
      link="scanner_body_link" />
    <child
      link="left_camera_link" />
    <axis
      xyz="0 0 0" />
  </joint>
  <link
    name="right_camera_link">
    <inertial>
      <origin
        xyz="2.7718E-06 8.7642E-07 0.0071874"
        rpy="0 0 0" />
      <mass
        value="0.0016829" />
      <inertia
        ixx="5.6994E-08"
        ixy="-6.0636E-12"
        ixz="5.3509E-12"
        iyy="5.6998E-08"
        iyz="-2.5605E-12"
        izz="2.9435E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/right_camera_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.79216 0.81961 0.93333 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://binocular_laser_profile_scanner/meshes/right_camera_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="right_camera_joint"
    type="fixed">
    <origin
      xyz="-0.073979 0.0175 0.031514"
      rpy="0 0.85197 0" />
    <parent
      link="scanner_body_link" />
    <child
      link="right_camera_link" />
    <axis
      xyz="0 0 0" />
  </joint>

  <!-- projector deneme v1 
  <gazebo reference="line_laser_link">
    <material>Gazebo/Red</material>
    <projector name="line_projector">
      <pose> 0 0 0 0 0 0 </pose>
      <texture>building_frame.png</texture>
      <near_clip>0.01</near_clip>
      <far_clip>10.0</far_clip>
      <fov>0.1</fov>
      <plugin name="projector_controller" filename="libgazebo_ros_projector.so">
        <update_rate>10.0</update_rate>
      </plugin>    
    </projector>
  </gazebo>
  -->

  <gazebo reference="line_laser_link">
    <projector name="projector">
      <pose>0 0 0 0 0 0</pose>
      <texture>building_frame.png</texture>
      <fov>1.0</fov>
      <near_clip>0.1</near_clip>
      <far_clip>10</far_clip>
    </projector>
  </gazebo>
  <gazebo>
    <plugin name="projector_controller" filename="libgazebo_ros_projector.so">
      <projector>line_laser_link/projector</projector>
      <alwaysOn>true</alwaysOn>
      <updateRate>15.0</updateRate>
      <textureName>building_frame.png</textureName>
      <textureTopicName>projector_controller/image</textureTopicName>
      <projectorTopicName>projector_controller/projector</projectorTopicName>
    </plugin>
  </gazebo>

  <!-- Line Laser Projector -->
  <!--
  <xacro:include filename="$(find binocular_laser_profile_scanner)/urdf/gazebo_projector.xacro"/>
  -->

</robot>